nonffi/writer.go

```diff
@@ -0,0 +1,191 @@
+package nonffi
+
+import (
+	"math/bits"
+	"runtime"
+	"sync"
+
+	"github.com/filecoin-project/go-commp-utils/zerocomm"
+	"github.com/filecoin-project/go-fil-commcid"
+	"github.com/filecoin-project/go-fil-commp-hashhash"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+type DataCIDSize struct {
+	PayloadSize int64
+	PieceSize   abi.PaddedPieceSize
+	PieceCID    cid.Cid
+}
+
+const commPBufPad = abi.PaddedPieceSize(8 << 20)
+const CommPBuf = abi.UnpaddedPieceSize(commPBufPad - (commPBufPad / 128)) // can't use .Unpadded() for const
+
+type ciderr struct {
+	c   cid.Cid
+	err error
+}
+
+type Writer struct {
+	len    int64
+	buf    [CommPBuf]byte
+	leaves []chan ciderr
+
+	tbufs    [][CommPBuf]byte
+	throttle chan int
+}
+
+func (w *Writer) Write(p []byte) (int, error) {
+	if w.throttle == nil {
```

Comment:
The manual throttling/error propagation mechanism can be replaced with this corelib:
```
	var eg errgroup.Group
	eg.SetLimit(runtime.NumCPU())

	for ... {
		eg.Go( func() error { ... } )
	}

	return eg.Wait()
```

Since you know how may "segments" you have, you can preallocate an array, and have each individual worker write out the commp bytes to the appropriate "cell" within that buffer.

This way you entirely obvioate the need of messge passing / channel synchronization / etc